### DISCUSSION TITLE
zero coordinate point of icub && inverse kinematics

### DISCUSSION THREAD

**Discussion by jimingda** [robotology/community#643] [Q&A] [answered]
The task i want to do is let the arm(hand) follow serial of positons (i.e spatial coordinate points), my problem is:
1. Where is the zero coordinate point of icub, so i can derive the desired spatial coordinate related to zero point.
2. Can i read the specific spatial coordinate of joint or terminal of icub on specific icub pose, maybe there is some api about it.
3. Are there any tutorials about this kind application, or maybe inverse kinematics are the most related one.

**Comment by pattacini** [comment, marked as answer]
Hi @jimingda 

>1. Where is the zero coordinate point of icub, so i can derive the desired spatial coordinate related to zero point.

The root reference frame is attached to the robot's waist and oriented as per https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-forward-kinematics/icub-forward-kinematics.

>2. Can i read the specific spatial coordinate of joint or terminal of icub on specific icub pose, maybe there is some api about it.

Sure! You can use the `ICartesianInterface`, whose documentation is available at https://robotology.github.io/robotology-documentation/doc/html/icub_cartesian_interface.html.

>3. Are there any tutorials about this kind application, or maybe inverse kinematics are the most related one.

https://github.com/vvv-school/tutorial_cartesian-interface is a good starting point.

**Reply by jimingda** [reply]
Thanks @pattacini   I also found a similar discussion below. Actually my task is driving a Logic G923 Wheel by icub hands, my firstly plan is to turn the steering wheel to the left by 15 degree,  i may need to generate the trajectory or get the positions directly under compliant mode, and then make two hand follow these positions, i am not sure whether it will need torso motion or not. After these, the second step is to use artificial skin, maybe make two hands grasp the wheel.
https://github.com/orgs/robotology/discussions/366

**Reply by pattacini** [reply]
It may be that IK won't perform quickly as well as precisely simultaneously for your task. To avoid losing performance because of such a trade-off, you may consider solving IK offline for a set of waypoints representing the problem at hand and then interpolating online among the solutions.

We did something similar for the [air hockey task](https://github.com/pattacini/study-air-hockey), which could be worth checking.

**Reply by jimingda** [reply]
Currently i came across a problem that when i want to open left and right arm together,  "Unable to open the Cartesian Controller"

```
  2  17.997565 <INFO> threadInit
  3  18.053926 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/left_arm") (remote "/icubSim/cartesianController/left_arm")
  4  18.317543 <INFO> Created wrapper <cartesiancontrollerclient>. See C++ class ClientCartesianController for documentation.
  5  18.317606 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
  6  18.652605 <ERROR> unable to connect to solver!
  7  19.616556 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
  8  28.619064 <ERROR> Unable to open the Cartesian Controller

```

```c++
class CtrlThread : public PeriodicThread
{
protected:
    PolyDriver        client_l, client_r;
    ICartesianControl *arm_l, *arm_r;

    Vector xd_l, xd_r;
    Vector od_l, od_r;

    int startup_context_id_l, startup_context_id_r;

    double t;
    double t0;
    double t1;

public:
  CtrlThread() : PeriodicThread(1.0) {}
  virtual bool threadInit()
  {
      yInfo() << "threadInit";
      Property option_l, option_r;
      option_l.put("device", "cartesiancontrollerclient");
      option_l.put("remote", "/icubSim/cartesianController/left_arm");
      option_l.put("local", "/cartesian_client/left_arm");

      option_r.put("device", "cartesiancontrollerclient");
      option_r.put("remote", "/icubSim/cartesianController/right_arm");
      option_r.put("local", "/cartesian_client/right_arm");

      // let's give the controller some time to warm up
      bool ok = false;
      t0 = Time::now();
      while (Time::now() - t0 < 10.0)
      {
          // this might fail if controller
          // is not connected to solver yet
          if (client_l.open(option_l) && client_r.open(option_r))
          {
              ok = true;
              break;
          }

          Time::delay(1.0);
      }
      if (!ok)
      {
          yError() << "Unable to open the Cartesian Controller";
          return false;
      }
```

**Reply by pattacini** [reply]
Hi @jimingda 

Line 6 of the log

>6  18.652605 <ERROR> unable to connect to solver!

tells us that one of the controllers is unable to connect to the corresponding solver!
Are the two instances of `iKinCartesianSolver` running ok?

Also, I've spotted the following line in your code

```c++
if (client_l.open(option_l) && client_r.open(option_r))
```

which is prone to errors as, once one of the controllers is connected, you shouldn't try to reconnect it again. The latter will happen as the two connections will most likely take place at different time instances.

**Reply by jimingda** [reply]
Thanks@pattacini . 
I modified the open method to seperatly open left and right arm

```
  virtual bool threadInit()
  {
      yInfo() << "threadInit";
      Property option_l, option_r;
      option_l.put("device", "cartesiancontrollerclient");
      option_l.put("remote", "/icubSim/cartesianController/left_arm");
      option_l.put("local", "/cartesian_client/left_arm");

      option_r.put("device", "cartesiancontrollerclient");
      option_r.put("remote", "/icubSim/cartesianController/right_arm");
      option_r.put("local", "/cartesian_client/right_arm");

      bool ok = false;
      t0 = Time::now();
      while (Time::now() - t0 < 10.0)
      {
          if (client_l.open(option_l))
          {
              ok = true;
              break;
          }

          Time::delay(1.0);
      }
      if (!ok)
      {
          yError() << "Unable to open the Cartesian Controller Left";
          return false;
      }

      ok = false;
	t0 = Time::now();
	while (Time::now() - t0 < 10.0)
	{
		if (client_r.open(option_r))
		{
			ok = true;
			break;
		}

		Time::delay(1.0);
	}

	if (!ok)
	{
		yError() << "Unable to open the Cartesian Controller Right";
		return false;
	}
```

after launch the app, unable to connect to solver! , Unable to open the Cartesian Controller Right.
```
 20.128260 <INFO> threadInit
 20.188709 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/left_arm") (remote "/icubSim/cartesianController/left_arm")
 20.476399 <INFO> Created wrapper <cartesiancontrollerclient>. See C++ class ClientCartesianController for documentation.
 20.476457 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
 20.696100 <ERROR> unable to connect to solver!
 21.298772 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
 22.303518 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
 22.409293 <ERROR> unable to connect to solver!
 23.005406 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
 24.001651 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
 24.401404 <ERROR> unable to connect to solver!
 25.251253 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
 26.246261 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
 26.463691 <ERROR> unable to connect to solver!
 27.021784 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
 28.021083 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
 28.375414 <ERROR> unable to connect to solver!
 29.064104 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
 30.059464 <DEBUG> Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/right_arm") (remote "/icubSim/cartesianController/right_arm")
 30.436635 <ERROR> unable to connect to solver!
 31.265806 <ERROR> Driver <cartesiancontrollerclient> was found but could not open
 32.263960 <ERROR> Unable to open the Cartesian Controller Right
```

besides, some errors in yarprobotinterface log, maybe it's the root cause that cann't connect to cartesianSolver/right_arm
```
 5.250769 <INFO> Created device <cartesiancontrollerserver>. See C++ class ServerCartesianController for documentation.
 5.250866 <INFO> part #1: left_arm
 5.250886 <INFO> joint #3: [-95, 10] deg
 5.250900 <INFO> joint #9: [-20, 25] deg
 5.260973 <INFO> icubSim/cartesianController/left_arm: Connecting to cartesian solver cartesianSolver/left_arm...
 5.321488 <INFO> icubSim/cartesianController/left_arm: Connections established with cartesianSolver/left_arm
 5.331777 <INFO> Executing attach action, level 6 on device right_arm_cartesian with parameters [("networks" = "(torso right_arm)"), ("torso" = "torso"), ("right_arm" = "right_arm")]
 5.331848 <INFO> Acquiring info on driver right_arm...
 5.331861 <INFO> icubSim/cartesianController/right_arm: IEncodersTimed interface will be used
 5.331873 <INFO> part #0: torso
 5.331886 <INFO> joint #0: [-20, 70] deg
 5.331895 <INFO> joint #1: [-30, 30] deg
 5.331904 <INFO> part #1: right_arm
 5.331917 <INFO> joint #3: [-95, 10] deg
 5.331926 <INFO> joint #4: [0, 160] deg
 5.331935 <INFO> joint #5: [-37, 80] deg
 5.331944 <INFO> joint #6: [15, 106] deg
 5.331953 <INFO> joint #7: [-60, 60] deg
 5.331961 <INFO> joint #8: [-80, 25] deg
 5.331969 <INFO> joint #9: [-20, 25] deg
 5.341840 <INFO> icubSim/cartesianController/right_arm: Checking if cartesian solver cartesianSolver/right_arm is alive... not yet
 5.341891 <INFO> Starting icubSim/cartesianController/right_arm at 10 ms
 5.341902 <INFO> All actions for action level 6 of startup phase finished.
 5.341912 <INFO> run phase starting...
 5.341921 <DEBUG> yarprobotinterface running happily
 6.359430 <INFO> icubSim/cartesianController/right_arm: Checking if cartesian solver cartesianSolver/right_arm is alive... not yet
 7.393769 <INFO> icubSim/cartesianController/right_arm: Checking if cartesian solver cartesianSolver/right_arm is alive... not yet
```

**Reply by jimingda** [reply]
I just found that the default parameter of iKinCartesianSolver is left_arm, right_arm was not specified, may i should add "--part right_arm"
![2023-07-15_16-07](https://github.com/robotology/community/assets/7019692/72f034a7-ba09-456a-848f-c3d484ad9a11)

**Reply by jimingda** [reply]
Another problem is that i cannt launch yarprobotinterface && iKinCartesianSolver of  tutorial_cartesian-interface system on the ubuntu, not like the docker above.
![2023-07-15_17-04](https://github.com/robotology/community/assets/7019692/5ccf2d34-81f2-4a01-a41c-d0ea85bc6cac)

**Reply by pattacini** [reply]
> I just found that the default parameter of iKinCartesianSolver is left_arm, right_arm was not specified, may i should add "--part right_arm"

Nope, you need 2 `iKinCartesianSolver` instances, one per arm, correctly configured.
See, for example, https://github.com/robotology/icub-gazebo-grasping-sandbox/blob/master/scripts/icub-grasp.xml#L32-L50.

>Another problem is that i cannt launch yarprobotinterface && iKinCartesianSolver of tutorial_cartesian-interface system on the ubuntu, not like the docker above.

From the screenshot, it appears that ports like `/icubSim/*` are not found, hence the simulator didn't get started as expected, probably.

Try launching those services straight from the console to get more insights from the printouts.

**Reply by jimingda** [reply]
I add the following part in tutorial_cartesian-interface/app/scripts/tutorial_cartesian-interface-system.xml, but it seems there is still one iKinCartesianSolver on the yarpmanager tutorial_cartesian-interface-system gui.
```
  <module>
      <name>iKinCartesianSolver</name>
      <parameters>--context gazeboCartesianControl --part right_arm</parameters>
      <dependencies>
          <port timeout="20">/yarplogger</port>
          <port timeout="20">/icubSim/torso/state:o</port>
          <port timeout="20">/icubSim/right_arm/state:o</port>
      </dependencies>
      <environment>YARP_FORWARD_LOG_ENABLE=1</environment>
      <node>localhost</node>  
      <ensure>
          <wait>2</wait>
      </ensure>
  </module>

```

**Reply by pattacini** [reply]
This simply means that either you didn't refresh the GUI or you edited the wrong XML file.

To open the XML file from the GUI straight away, follow these [instructions](https://github.com/orgs/robotology/discussions/638#discussioncomment-6251730).
You could set up your system to use for example `gedit` to open XML files instead of the default app, which is usually the browser.

**Reply by jimingda** [reply]
Thanks@pattachi  
After i modify the correct xml the iKinCartesianSolver for right_arm could be created. And is there any api that can check if the arm end reach some specific position?

**Reply by jimingda** [reply]
Another thing i found is that it seems there isn't collision detection since the two arms(hands) can be overlapped on the simulation.

**Reply by pattacini** [reply]
>After i modify the correct xml the iKinCartesianSolver for right_arm could be created.

Happy it worked!

>And is there any api that can check if the arm end reach some specific position?

Sure, it's the same API documented at https://robotology.github.io/robotology-documentation/doc/html/icub_cartesian_interface.html.

In particular, you can peruse:
- [**`checkMotionDone`**](https://yarp.it/latest/classyarp_1_1dev_1_1ICartesianControl.html#abb6a6f62d165169c66e22fb8e749d417)
- [**`waitMotionDone`**](https://yarp.it/latest/classyarp_1_1dev_1_1ICartesianControl.html#af1a253cb67ed5cf091854a58248e6446)

>Another thing i found is that it seems there isn't collision detection since the two arms(hands) can be overlapped on the simulation.

You're right, there's no collision avoidance implemented between the two arms.
You ought to implement something yourself at a higher level.

**Reply by jimingda** [reply]
As the suggestion above "Try launching those services straight from the console to get more insights from the printouts." I follow the following steps after launch yarpserver.
```
      // 1 - the iCub simulator is running
      //     (launch: iCub_SIM)
      // 2 - the cartesian server is running
      //     (launch: yarprobotinterface --context simCartesianControl)
      // 3 - the cartesian solver for the left arm is running too
      //     (launch: iKinCartesianSolver --context simCartesianControl --part left_arm)
```
It seems yarprobotinterface didn't work well with follow information nonstop
```
[INFO] icubSim/cartesianController/right_arm: Checking if cartesian solver cartesianSolver/right_arm is alive... not yet
[INFO] icubSim/cartesianController/left_leg: Checking if cartesian solver cartesianSolver/left_leg is alive... not yet
[INFO] icubSim/cartesianController/right_leg: Checking if cartesian solver cartesianSolver/right_leg is alive... not yet
[INFO] icubSim/cartesianController/right_arm: Checking if cartesian solver cartesianSolver/right_arm is alive... not yet
[INFO] icubSim/cartesianController/left_leg: Checking if cartesian solver cartesianSolver/left_leg is alive... not yet
```

**Reply by jimingda** [reply]
Maybe i should use the right command with the correct parameters as below. I will try it with simCartesianControl
```
yarprobotinterface --context gazeboCartesianControl --config no_legs.xml
iKinCartesianSolver --context gazeboCartesianControl --part left_arm
```

**Reply by pattacini** [reply]
Yep, I've cleaned up the [comments in the code](https://github.com/vvv-school/tutorial_cartesian-interface/commit/bcbadbe282256b2577c15542607b4d260563770f) as they were referring to some old resources. 

You ought to launch the same commands you see in the `yarpmanager`, with their exact options.

**Reply by jimingda** [reply]
My another question is that why spawn a PeriodicThread to implement tutorial_cartesian-interface rather than do it in the updateModule, like other applications?

**Reply by pattacini** [reply]
The class `yarp::os::PeriodicThread` provides a more precise timing than `yarp::os::RFModule`, especially when you require faster rates. You can still use the latter for background non-demanding tasks.

**Reply by pattacini** [reply]
Hi @jimingda 

Did you manage to address all the points?

**Reply by jimingda** [reply]
Yes. I can address the left and right hand position, but still don't know the relation of position of finger to the hand position.
I find the discussion before  https://github.com/orgs/robotology/discussions/398 . My current plan is to send position to finger joints(thumb proximal,  thumb distal, index proximal, index distal, middle proximal, middle distal) until the finger tactile skin touch the target(wheel).

**Reply by pattacini** [reply]
>My current plan is to send position to finger joints(thumb proximal, thumb distal, index proximal, index distal, middle proximal, middle distal) until the finger tactile skin touch the target(wheel).

I deem this to be somewhat unrelated to this particular discussion.
Please, consider marking one post as answer, if you're confident that the thread is complete.

For any other matter regarding finger control, feel free to comment on #398 or to open a fresh new discussion.
